horovod/common/controller.cc

```diff
@@ -961,17 +992,18 @@ void Controller::FuseResponses(std::deque<Response>& responses,
         const auto& new_entry =
             tensor_queue_.GetTensorEntry(new_response.tensor_names()[0]);
 
-        int64_t new_total_byte_size_of_output = TotalByteSizeOfAllgatherOutput(
+        auto new_tensor_byte_sizes = ByteSizeOfEachAllgatheredTensor(
```

Comment:
This is creating a new vector every iteration of this loop, which may increase latency. Do you think it'd be worth it to try to avoid it?

Reply:
Very good point. Avoiding latency is important here. 

I've changed this in https://github.com/horovod/horovod/pull/3727/commits/df6c4385dab5c5c7784f9e938d20528b14f34ab3 so the memory for those two allgather size vectors is allocated just once at the start of `FuseResponses()`. A slight drawback might be that the control flow in this long function is less local now, making it a bit easier to introduce bugs.

Reply:
Thanks for updating this! I think avoiding the allocations is well worth the added complexity of the implementation, given that `FuseResponses` is on the critical path and always called, even in cases where the responses are cached.